### PR TITLE
Add HKElectrocardiogram to Observation Mapping

### PR DESCRIPTION
**Pull request by PSchmiedmayer**
# Add HKElectrocardiogram to Observation Mapping

## :recycle: Current situation & Problem
`HKElectrocardiogram` is a commonly used HealthKit sample type that is collected using an Apple Watch or other ECG devices. 

## :bulb: Proposed solution
Adds an mapping of an ecg to an FHIR observation.
The observation uses this format as displayed in the example below.

The `component` section includes the Apple provided `HKElectrocardiogram.NumberOfVoltageMeasurements`, `HKElectrocardiogram.SamplingFrequency`, and `HKElectrocardiogram.Classification` metadata associated with an ECG.

The `component` section also includes multiple `HKCategoryType` encoded values representing the symptoms if they are passed in the `observation(symptoms: Symptoms, voltageMeasurements: VoltageMeasurements, withMapping mapping: HKSampleMapping)` function.

The `component` section also includes multiple `valueSampledData` raw voltage measurements if they are passed into the `observation(symptoms: Symptoms, voltageMeasurements: VoltageMeasurements, withMapping mapping: HKSampleMapping)` function. The raw voltage measurements are divided up into 10 second pieces.
The `valueSampledData` instances follow the FHIR specification: https://build.fhir.org/datatypes.html#sampleddata .

```json
{
  "category" : [
    {
      "coding" : [
        {
          "code" : "procedure",
          "display" : "Procedure",
          "system" : "http://terminology.hl7.org/CodeSystem/observation-category"
        }
      ]
    }
  ],
  "code" : {
    "coding" : [
      {
        "code" : "HKElectrocardiogram",
        "display" : "Electrocardiogram",
        "system" : "http://developer.apple.com/documentation/healthkit"
      },
      {
        "code" : "131329",
        "display" : "MDC_ECG_ELEC_POTL_I",
        "system" : "urn:oid:2.16.840.1.113883.6.24"
      }
    ]
  },
  "component" : [
    {
      "code" : {
        "coding" : [
          {
            "code" : "HKElectrocardiogram.NumberOfVoltageMeasurements",
            "display" : "Electrocardiogram Number of Voltage Measurements",
            "system" : "http://developer.apple.com/documentation/healthkit"
          }
        ]
      },
      "valueQuantity" : {
        "unit" : "measurements",
        "value" : 22112
      }
    },
    {
      "code" : {
        "coding" : [
          {
            "code" : "HKElectrocardiogram.SamplingFrequency",
            "display" : "Sampling Frequency",
            "system" : "http://developer.apple.com/documentation/healthkit"
          }
        ]
      },
      "valueQuantity" : {
        "code" : "hertz",
        "system" : "http://unitsofmeasure.org",
        "unit" : "hertz",
        "value" : 512
      }
    },
    {
      "code" : {
        "coding" : [
          {
            "code" : "HKElectrocardiogram.Classification",
            "display" : "Electrocardiogram Classification",
            "system" : "http://developer.apple.com/documentation/healthkit"
          }
        ]
      },
      "valueString" : "sinusRhythm"
    },
    {
      "code" : {
        "coding" : [
          {
            "code" : "8867-4",
            "display" : "Heart rate",
            "system" : "http://loinc.org"
          },
          {
            "code" : "HKQuantityTypeIdentifierHeartRate",
            "display" : "Heart Rate",
            "system" : "http://developer.apple.com/documentation/healthkit"
          }
        ]
      },
      "valueQuantity" : {
        "code" : "/min",
        "system" : "http://unitsofmeasure.org",
        "unit" : "beats/minute",
        "value" : 140
      }
    },
    {
      "code" : {
        "coding" : [
          {
            "code" : "HKElectrocardiogram.SymptomsStatus",
            "display" : "Electrocardiogram Symptoms Status",
            "system" : "http://developer.apple.com/documentation/healthkit"
          }
        ]
      },
      "valueString" : "present"
    },
    {
      "code" : {
        "coding" : [
          {
            "code" : "HKCategoryTypeIdentifierRapidPoundingOrFlutteringHeartbeat",
            "display" : "Rapid Pounding Or Fluttering Heartbeat",
            "system" : "http://developer.apple.com/documentation/healthkit"
          }
        ]
      },
      "valueString" : "unspecified"
    },
    {
      "code" : {
        "coding" : [
          {
            "code" : "HKCategoryTypeIdentifierFatigue",
            "display" : "Fatigue",
            "system" : "http://developer.apple.com/documentation/healthkit"
          }
        ]
      },
      "valueString" : "unspecified"
    },
    {
      "code" : {
        "coding" : [
          {
            "code" : "131329",
            "display" : "MDC_ECG_ELEC_POTL_I",
            "system" : "urn:oid:2.16.840.1.113883.6.24"
          }
        ]
      },
      "valueSampledData" : {
        "data" : "-122.59088134765625 -262.6523132324219 -398.10791015625 -517.7991333007812 ....",
        "dimensions" : 1,
        "origin" : {
          "code" : "uV",
          "system" : "http://unitsofmeasure.org",
          "unit" : "uV",
          "value" : 0
        },
        "period" : 0.00195303667126447232
      }
    }
  ],
  "effectivePeriod" : {
    "end" : "2022-12-20T01:41:00-08:00",
    "start" : "2022-12-20T01:41:43.1875-08:00"
  },
  "identifier" : [
    {
      "id" : "E3E98DA2-7124-4567-93CB-0B52647EA332"
    }
  ],
  "issued" : "2022-12-20T19:14:52.072656035-08:00",
  "resourceType" : "Observation",
  "status" : "final"
}
```

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PSchmiedmayer** [comment]
@vishnuravi This PR introduces a first draft to encode an ECG in a FHIR observation. 
I have decided to encode the ECG measurement using a similar approach to how we want to encode `HKCorrelation`s and workouts in the future. The configuration relies on `HKCategorySamples` as much as possible and then introduces `ECG` specific fields in the configuration.

I have tried encoding the raw voltage values but this generally leads to an about 6 MB large observation so I have decided to allow a developer to pass in an URL that is encoded in the voltage measurement field to direct to a storage URL where the raw measurements are encoded.

It would be great to get your idea of the approaches chosen in this PR. Once we agree on the overall approach I can add more tests un the UI tests and can clean up the code 👍

File: Sources/HealthKitOnFHIR/Resources/HKSampleMapping.json [hunk 1]
```diff
@@ -1,4 +1,93 @@
 {
+    "HKElectrocardiogram": {
+      "categories": [
+        {
+          "code": "vital-signs",
+          "display": "Vital Signs",
+          "system": "http://terminology.hl7.org/CodeSystem/observation-category"
```
Comment (vishnuravi) [inline Sources/HealthKitOnFHIR/Resources/HKSampleMapping.json:7]: The category code for an ECG should be "procedure", as noted in the FHIR documentation [here](https://www.hl7.org/fhir/codesystem-observation-category.html).

```suggestion
          "code": "procedure",
          "display": "Procedure",
          "system": "http://terminology.hl7.org/CodeSystem/observation-category"
```

**Comment by vishnuravi** [comment]
Hey @PSchmiedmayer! Thanks for working on this, this is great!

I think we should try to stay as close to the [ECG Observation example](https://fhir-ru.github.io/observation-example-sample-data.json.html) given in the FHIR documentation in order for this data to be readable by other systems.

This would entail coding the voltage measurements in components of the Observation as [SampledData](https://hl7.org/fhir/datatypes.html#SampledData) with a code representing the lead (as in this example for lead I): 

```
{
   "code": {
      "coding": [
         {
            "code": "131329",
            "system": "urn:oid:2.16.840.1.113883.6.24",
            "display": "MDC_ECG_ELEC_POTL_I"
         },
         {
            "display": "microvolt",
            "code": "uV",
            "system": "http://unitsofmeasure.org"
         }
      ]
   },
   "valueSampledData" : {
      "period": 1.953125,
      "dimensions": 1,
      "data": "**<Voltage Measurements Separated by Space Characters>**",
      "origin": {
         "value": 0
      }
   }
}
```
The `period` here represents the number of milliseconds between samples and `dimensions`, the number of samples at each time point.

Regarding the amount of data, a potential solution is to split the samples up into separate observations. (A standard ECG is a 10-second strip.) I will investigate this a bit more.

**Comment by PSchmiedmayer** [comment]
@vishnuravi I have updated the PR based on your feedback and improved the FHIR structure. I have used https://build.fhir.org/datatypes.html#sampleddata as a reference for the sample data. It seems like the spec requests an `interval` and an `intervalUnit` that is not exposed in the FHIRModels package. The package only includes the `period` key but specifies that it is the `"Number of milliseconds between samples"` which I calculate to be around 0.00195303667126447232 (aligns with 512 herz specified by Apple). Am I missing something here or is this an error in the FHIRModels Package not correctly reflecting the FHIR spec?

**Comment by vishnuravi** [comment]
> @vishnuravi I have updated the PR based on your feedback and improved the FHIR structure. I have used https://build.fhir.org/datatypes.html#sampleddata as a reference for the sample data. It seems like the spec requests an `interval` and an `intervalUnit` that is not exposed in the FHIRModels package. The package only includes the `period` key but specifies that it is the `"Number of milliseconds between samples"` which I calculate to be around 0.00195303667126447232 (aligns with 512 herz specified by Apple). Am I missing something here or is this an error in the FHIRModels Package not correctly reflecting the FHIR spec?

@PSchmiedmayer you are actually looking at the FHIR R5 spec there, the R4 spec does not have those two elements: https://hl7.org/fhir/datatypes.html#SampledData.

**Comment by vishnuravi** [comment]
Looks good to me! Thanks so much @PSchmiedmayer!

**Review by vishnuravi — APPROVED** [review]

**Comment by PSchmiedmayer** [comment]
Thank you @vishnuravi, I looked at the wrong version of the spec, that makes sense! Glad to hear that you agree with the implementation. I will add additional unit tests and then merge the PR 👍

**Pull request merged**